### DISCUSSION TITLE
Have issues with the typings making `update` argument not to infer in callback for some reason

### DISCUSSION THREAD

**Discussion by RIP21** [pmndrs/jotai#433] [Q&A] [answered]
I have the following weird behavior. On version ~0.12.5 it was working just fine, after an update to 0.16.1 got this issue below.
![image](https://user-images.githubusercontent.com/3940079/115160433-62eb6800-a098-11eb-8ef1-c614e1f7ecff.png)

It happening no matter if I'm using helper or `useAtom` itself.

Here is the code of the family atom itself.

```
export interface ProfitLossEngineState {
  id: string
  queries: ProfitLossQueries
  queryResult: {
    [key: string]: UseQueryState<
      NewProfitLossTable_ProfitLossQuery,
      NewProfitLossTable_ProfitLossQueryVariables
    >
  }
}

export interface ProfitLossFamilyInitialize {
  id: string
}

export const profitLossEngineAtom = atomFamily<
  ProfitLossFamilyInitialize,
  ProfitLossEngineState,
  ProfitLossEngineState
>(
  ({ id }) => {
    const anAtom = atom({ id, queries: {}, queryResult: {} })
    anAtom.onMount = (setAtom) => {
      setAtom({ id, queries: {}, queryResult: {} })
    }
    return anAtom
  },
  (a, b) => a.id === b.id,
)
```

The only way to get away with this one is to not use a callback but use an object spread which may be dangerous the same as with `useState` `this.setState` as it's async?

**Comment by dai-shi** [comment, marked as answer]
Thanks for opening this. We changed `atom()` typing in v0.16.1 hoping to "improve" it...
Does it work with v0.16.0?
I'd like to add a test to investigate it. Can you help reproducing this typing issue minimally? It can repro with useAtom without atomFamily, or not?

**Reply by RIP21** [reply]
@dai-shi here is the repro for 0.16.1
https://codesandbox.io/s/broken-typings-for-family-atom-setter-sezf1

In 0.16.0 it's broken for such a scenario too.
Although a few others disappeared :D (also included into repro code on line 62 App.tsx)

**Reply by dai-shi** [reply]
Thanks for the repro. They are two separate issues.

1. v0.16.1 introduces `NonFunction` which seems proper in types, but not good in DX. We'd like to improve it.
2. atomFamily's third type param should be SetStateAction<T> in your use case. It's easier to have it inferred.

Fixed version: https://codesandbox.io/s/broken-typings-for-family-atom-setter-forked-6w5jf?file=/src/App.tsx

**Reply by RIP21** [reply]
@dai-shi so to reiterate.
1. Got it
2. So I should either pass 3rd generic argument to be SetStateAction like that:
```ts
export const profitLossEngineAtom = atomFamily<
  ProfitLossFamilyInitialize,
  ProfitLossEngineState,
  SetStateAction<ProfitLossEngineState>
>(
  ({ id }) => {
    const anAtom = atom({ id, queries: {}, queryResult: {} });
    anAtom.onMount = (setAtom) => {
      setAtom({ id, queries: {}, queryResult: {} });
    };
    return anAtom;
  },
  (a, b) => a.id === b.id
);
```
or just explicitly type my first `atomFamily` with Init type explicitly and `jotai` will infer the rest.

**Reply by RIP21** [reply]
If that's the case, I think generic like
`atomFamily<Param, Value, Update extends SetStateAction<Value>>` may be better? :) 
I may be wrong surely as typings are hard :) But just a guess :)
Either way, thanks for an answer will keep that stuff above as an improvement task for you or whoever would like to contribute :)

**Reply by dai-shi** [reply]
> will keep that stuff above as an improvement task for you or whoever would like to contribute :)

Yeah. May I ask you to create an issue for it?

> `atomFamily<Param, Value, Update extends SetStateAction<Value>>`

This is not correct. (or only correct for primitive atoms.)
`Update` can be any type, not tied to `Value`.

For example:
```ts
const myFamily = atomFamily<string, string, number>((id: string) => atom(
  (get) => 'foo' + ...
  (get, set, update: number) => ...
))
```

Other than that, you got all correctly.

**Reply by RIP21** [reply]
@dai-shi done.
https://github.com/pmndrs/jotai/issues/436
Feel free to edit as I was in hurry :)

**Reply by dai-shi** [reply]
Oops,

> will keep that stuff above as an improvement task for you or whoever would like to contribute :)

I might have misunderstood what was "that stuff above". I meant about `NonFunction` stuff. Anyway, we can address both.
